distributed/worker.py

```diff
@@ -1555,10 +1586,104 @@ def transition(self, ts, finish, **kwargs):
             self.validate_task(ts)
         self._notify_plugins("transition", ts.key, start, state or finish, **kwargs)
 
-    def transition_waiting_flight(self, ts, worker=None):
+    def transition_new_waiting(self, ts):
         try:
             if self.validate:
-                assert ts.state != "flight"
+                assert ts.state == "new"
+                assert ts.runspec is not None
+                assert not ts.who_has
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb
+
+                pdb.set_trace()
+            raise
+
+    def transition_new_fetch(self, ts):
+        try:
+            if self.validate:
+                assert ts.state == "new"
+                assert ts.runspec is None
+
+            for dependent in ts.dependents:
+                dependent.waiting_for_data.add(ts.key)
+
+            self.data_needed.append(ts.key)
```

Comment:
I noticed the `data_needed` to be occasionally populated by duplicates. Since it is a list, we cannot really guarantee anything else. I believe this is not really harmful but we might want to consider a `key in data_needed` check